File: python/백준/스택, 큐, 덱/4949 균형잡힌 세상.py
while True:
    string = input()
    
    stack = []
    
    if string == '.':
        break
    
    for char in string:
        if char == "(":
            stack.append("(")
        elif char == "[":
            stack.append("[")
        
        elif char == ")":
            if len(stack) != 0 and stack[-1] == "(":
                stack.pop()
            else:
                stack.append(")")
                break
        elif char == "]":
            if len(stack) != 0 and stack[-1] == "[":
                stack.pop()
            else:
                stack.append("]")
                break
        else:
            continue
        
    if len(stack) == 0:
        print("yes")
    else:
        print("no")

File: cpp/백준/자료구조/스택, 집합/4949 균형잡힌 세상.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    string line;
    while (getline(cin, line)) {
        if (line == ".") break;

        stack<char> st;
        bool broken = false;

        for (char c : line) {
            if (c == '(' || c == '[') {
                st.push(c);
            } else if (c == ')') {
                if (!st.empty() && st.top() == '(') {
                    st.pop();
                } else {
                    broken = true;
                    break;
                }
            } else if (c == ']') {
                if (!st.empty() && st.top() == '[') {
                    st.pop();
                } else {
                    broken = true;
                    break;
                }
            }
        }

        if (!broken && st.empty()) cout << "yes\n";
        else cout << "no\n";
    }
    return 0;
}
